### PR TITLE
fix detection of functions in presence of indirections

### LINKED ISSUE #505: bug in operator creation when array only used as index
```
from devito import Grid, Function, Operator, Eq, Dimension
import numpy as np

grid = Grid(shape=(10,10))
x, y = grid.dimensions

p_poke = Dimension('p_src')
d = Dimension('d')

npoke = 1

u = Function(name='u', grid=grid, space_order=0) 

coordinates = Function(name='coordinates',
    dimensions=(p_poke, d),
    shape=(npoke, grid.dim), space_order=0, dtype=np.int32)

coordinates.data[0,0] = 4
coordinates.data[0,1] = 3

poke_eq = Eq(u.indexed[coordinates.indexed[p_poke,0],
                       coordinates.indexed[p_poke,1]], 1.0)

op = Operator(poke_eq)
print(op)

# The need for the p_src_e=2 is, I think, another issue
op.apply(p_src_e=2)

print(u.data)
```

Errors with:

```
Traceback (most recent call last):
  File "indexed_as_index_bug.py", line 24, in <module>
    print(op)
  File ".../devito/ir/iet/nodes.py", line 103, in __str__
    return str(self.ccode)
  File ".../devito/ir/iet/nodes.py", line 77, in ccode
    return CGen().visit(self)
  File ".../devito/tools.py", line 544, in visit
    return meth(o, *args, **kwargs)
  File ".../devito/ir/iet/visitors.py", line 290, in visit_Operator
    decls = self._args_decl(params)
  File ".../devito/ir/iet/visitors.py", line 146, in _args_decl
    if i.is_Object:
AttributeError: 'Symbol' object has no attribute 'is_Object'
```

It all starts to work if I use the coordinates array in a numerical expression, rather than only as an index.

e.g.:
```
-poke_eq = Eq(u.indexed[coordinates.indexed[p_poke,0], coordinates.indexed[p_poke,1]], 1.0)
+poke_eq = Eq(u.indexed[coordinates.indexed[p_poke,0], coordinates.indexed[p_poke,1]], coordinates.indexed[p_poke,0]*1e-20 + 1.0)
```

giving, as desired:
```
[[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 1. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]]
```

### LINKED ISSUE #513: Automatic detection of flow with mixed save/nosave time functions
So a small bug that turned out trying to save one of the fields in `born`, 

```
from devito import *
grid = Grid((10, 10))
u1 = TimeFunction(name="u1", grid=grid, save=10, time_order=1)
u2 = TimeFunction(name="u2", grid=grid, time_order=2)
u1 = TimeFunction(name="u1", grid=grid, save=10, time_order=2)
op = Operator([Eq(u1.forward, u1 + 2.0 - u1.backward), Eq(u2.forward, u2 + 2*u2.backward -  u1.dt2)])
```

Fails to detect the flow direction

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/nethome/mlouboutin3/Dropbox/London/CodeGen/devito/devito/backends.py", line 75, in __call__
    return t(*args, **kwargs)
  File "/nethome/mlouboutin3/Dropbox/London/CodeGen/devito/devito/core/operator.py", line 54, in __new__
    obj.__init__(*args, **kwargs)
  File "/nethome/mlouboutin3/Dropbox/London/CodeGen/devito/devito/operator.py", line 82, in __init__
    clusters = clusterize(expressions)
  File "/nethome/mlouboutin3/Dropbox/London/CodeGen/devito/devito/ir/clusters/algorithms.py", line 232, in clusterize
    flowmap = detect_flow_directions(g)
  File "/nethome/mlouboutin3/Dropbox/London/CodeGen/devito/devito/ir/support/utils.py", line 98, in detect_flow_directions
    for k, v in mapper.items() if k.is_Stepping)
AssertionError
```

### PR DESCRIPTION
**Pull request by FabioLuporini**
depends on #512 

fixes #505 

took the opportunity for some refactoring

EDIT: now also

fixes #513 

(@tjb900 @mloubout these were bugs you raised; not sure if Travis will be green because I'm at the airport on battery and I've only run a quick sample of them)

### CODE REVIEW

File: devito/ir/support/utils.py [hunk 1]
```diff
@@ -86,16 +86,9 @@ def detect_flow_directions(exprs):
     mapper.update({d: {Any} for d in flatten(i.aindices for i in reads + writes)
                    if d is not None and d not in mapper})
 
-    # Add in stepping dimensions (just in case they haven't been detected yet)
-    # note: stepping dimensions may force a direction on the parent
-    assert all(v == {Any} or mapper.get(k.parent, v) in [v, {Any}]
-               for k, v in mapper.items() if k.is_Stepping)
+    # Add in derived-dimensions parents, in case they haven't been detected yet
```
Comment (mloubout) [inline devito/ir/support/utils.py]: Couldn't put it above. Line 82:
`for d in dimensions[dimensions.index(d) + 1:]`
this fails as `d` is referenced before initialization.

Comment (FabioLuporini) [reply devito/ir/support/utils.py]: Fails where/when? It should pick the last d from the previus loop

Comment (mloubout) [reply devito/ir/support/utils.py]: When `dimensions` is empty

Comment (mloubout) [reply devito/ir/support/utils.py]: There is few other issue with `detect_flow_directions` I'm trying to create example for

Comment (mloubout) [reply devito/ir/support/utils.py]: I'm actually not sure wht this two lines (82-83) do as all `d` in `dimension` is added to the mapper in the previous loop.

Comment (FabioLuporini) [reply devito/ir/support/utils.py]: > When dimensions is empty

good spot

Comment (FabioLuporini) [reply devito/ir/support/utils.py]: > I'm actually not sure wht this two lines (82-83) do as all d in dimension is added to the mapper in the previous loop.

No, in the previous loop you ``break`` ASAP

**Comment by FabioLuporini** [comment]
* Rebased on ``master``
* Addressed @mloubout 's comments
* Now hopefully with a green travis

Probably mergeable if Travis gets green

**Pull request merged**